kythe/go/serving/pipeline/nodes/nodes.go

```diff
@@ -0,0 +1,193 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package nodes provides Beam transformations over *ppb.Nodes.
+package nodes
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+
+	"kythe.io/kythe/go/services/graphstore/compare"
+	"kythe.io/kythe/go/util/schema"
+	"kythe.io/kythe/go/util/schema/edges"
+	"kythe.io/kythe/go/util/schema/facts"
+
+	"github.com/apache/beam/sdks/go/pkg/beam"
+
+	ppb "kythe.io/kythe/proto/pipeline_go_proto"
+	scpb "kythe.io/kythe/proto/schema_go_proto"
+	spb "kythe.io/kythe/proto/storage_go_proto"
+)
+
+func init() {
+	beam.RegisterFunction(embedSourceKey)
+	beam.RegisterFunction(entryToNode)
+
+	beam.RegisterType(reflect.TypeOf((*combineNodes)(nil)).Elem())
+}
+
+// FromEntries transforms a PCollection of *ppb.Entry protos into *ppb.Nodes.
+func FromEntries(s beam.Scope, entries beam.PCollection) beam.PCollection {
+	return beam.ParDo(s, embedSourceKey,
+		beam.CombinePerKey(s, &combineNodes{},
+			beam.ParDo(s, entryToNode, entries)))
+}
+
+func entryToNode(e *spb.Entry, emit func(*spb.VName, *ppb.Node)) error {
+	if e.Source == nil {
+		return fmt.Errorf("invalid Entry: source is missing: %+v", e)
+	}
+
+	n := &ppb.Node{}
+	if e.EdgeKind == "" {
+		if e.FactName == "" || e.Target != nil {
+			return fmt.Errorf("invalid fact Entry: {%v}", e)
+		}
+
+		switch e.FactName {
+		case facts.NodeKind:
+			kind := string(e.FactValue)
+			if k := schema.NodeKind(kind); k != scpb.NodeKind_UNKNOWN_NODE_KIND {
+				n.Kind = &ppb.Node_KytheKind{k}
+			} else {
+				n.Kind = &ppb.Node_GenericKind{kind}
+			}
+		case facts.Subkind:
+			subkind := string(e.FactValue)
+			if k := schema.Subkind(subkind); k != scpb.Subkind_UNKNOWN_SUBKIND {
+				n.Subkind = &ppb.Node_KytheSubkind{k}
+			} else {
+				n.Subkind = &ppb.Node_GenericSubkind{subkind}
+			}
+		default:
+			n.Fact = append(n.Fact, entryToFact(e))
+		}
+	} else {
+		if (e.FactName != "/" && e.FactName != "") || len(e.FactValue) != 0 || e.Target == nil {
+			return fmt.Errorf("invalid edge Entry: {%v}", e)
+		}
+
+		n.Edge = append(n.Edge, entryToEdge(e))
+	}
+
+	emit(e.Source, n)
+	return nil
+}
+
+func entryToEdge(e *spb.Entry) *ppb.Edge {
+	kind, ord, _ := edges.ParseOrdinal(e.EdgeKind)
+	g := &ppb.Edge{Target: e.Target, Ordinal: int32(ord)}
+	edgeKind := schema.EdgeKind(kind)
+	if edgeKind == scpb.EdgeKind_UNKNOWN_EDGE_KIND {
+		g.Kind = &ppb.Edge_GenericKind{kind}
+	} else {
+		g.Kind = &ppb.Edge_KytheKind{edgeKind}
+	}
+	return g
+}
+
+func entryToFact(e *spb.Entry) *ppb.Fact {
+	f := &ppb.Fact{Value: e.FactValue}
+	name := schema.FactName(e.FactName)
+	if name == scpb.FactName_UNKNOWN_FACT_NAME {
+		f.Name = &ppb.Fact_GenericName{e.FactName}
+	} else {
+		f.Name = &ppb.Fact_KytheName{name}
+	}
+	return f
+}
+
+type combineNodes struct{}
+
+func (combineNodes) CreateAccumulator() *ppb.Node { return &ppb.Node{} }
```

Comment:
I don't recognize this declaration syntax: what is `combineNodes` here, since it appears not to be in the form of a receiver?  I mean, since in this case you don't need a variable to refer to it in the method body, I can see why Go might let you skip that part of the declaration, but if there's something else going on here, I'd like to understand it.

Reply:
Yes, it is a required combiner method.